pkg/urbit/vere/king.c

```diff
@@ -231,54 +236,161 @@ _king_curl_alloc(void* dat_v, size_t uni_t, size_t mem_t, void* buf_v)
     u3l_log("failed to initialize libcurl\n");
     exit(1);
   }
 
-  u3K.ssl_curl_f(curl);
-  curl_easy_setopt(curl, CURLOPT_URL, url_c);
-  curl_easy_setopt(curl, CURLOPT_WRITEFUNCTION, _king_curl_alloc);
-  curl_easy_setopt(curl, CURLOPT_WRITEDATA, (void*)&buf_u);
+  u3K.ssl_curl_f(cul_u);
+  curl_easy_setopt(cul_u, CURLOPT_URL, url_c);
+  curl_easy_setopt(cul_u, CURLOPT_WRITEFUNCTION, _king_curl_alloc);
+  curl_easy_setopt(cul_u, CURLOPT_WRITEDATA, (void*)&buf_u);
 
-  result = curl_easy_perform(curl);
-  curl_easy_getinfo(curl, CURLINFO_RESPONSE_CODE, &cod_l);
+  res_i = curl_easy_perform(cul_u);
+  curl_easy_getinfo(cul_u, CURLINFO_RESPONSE_CODE, &cod_i);
 
   //  XX retry?
   //
-  if ( CURLE_OK != result ) {
-    u3l_log("failed to fetch %s: %s\n",
-            url_c, curl_easy_strerror(result));
-    u3_king_bail();
-    exit(1);
+  if ( CURLE_OK != res_i ) {
+    u3l_log("curl: failed %s: %s\n", url_c, curl_easy_strerror(res_i));
+    ret_i = -1;
+  }
+  if ( 300 <= cod_i ) {
+    u3l_log("curl: error %s: HTTP %ld\n", url_c, cod_i);
+    ret_i = -2;
   }
-  if ( 300 <= cod_l ) {
-    u3l_log("error fetching %s: HTTP %ld\n", url_c, cod_l);
+
+  curl_easy_cleanup(cul_u);
+
+  *len_w = buf_u.len;
+  *hun_y = (c3_y*)buf_u.base;
+
+  return ret_i;
+}
+
+/* _king_get_atom(): HTTP GET url_c, produce response body as atom.
+*/
+static u3_noun
+_king_get_atom(c3_c* url_c)
+{
+  c3_w  len_w;
+  c3_y* hun_y;
+  u3_noun pro;
+
+  if ( _king_curl_bytes(url_c, &len_w, &hun_y) ) {
     u3_king_bail();
     exit(1);
   }
 
-  curl_easy_cleanup(curl);
+  pro = u3i_bytes(len_w, hun_y);
+  c3_free(hun_y);
+  return pro;
+}
+
+/* _king_get_pace(): get "pace" (release channel name).
+*/
+static c3_c*
+_king_get_pace(void)
+{
+  struct stat buf_u;
+  c3_y*       hun_y;
+  c3_c*       pat_c;
+  c3_w red_w, len_w;
+  c3_i ret_i, fid_i;
 
-  {
-    u3_noun pro = u3i_bytes(buf_u.len, (const c3_y*)buf_u.base);
+  ret_i = asprintf(&pat_c, "%s/.bin/pace", u3_Host.dir_c);
+  c3_assert( ret_i > 0 );
+
+  fid_i = c3_open(pat_c, O_RDONLY, 0644);
+
+  if ( (fid_i < 0) || (fstat(fid_i, &buf_u) < 0) ) {
+    c3_free(pat_c);
+    return strdup("live");
+  }
+
+  len_w = buf_u.st_size;
+  hun_y = c3_malloc(len_w);
+  red_w = read(fid_i, hun_y, len_w);
+  close(fid_i);
+
+  if ( len_w != red_w ) {
+    c3_free(pat_c);
+    c3_free(hun_y);
+    return strdup("live");
+  }
+
+  c3_free(pat_c);
 
-    c3_free(buf_u.base);
+  ret_i = asprintf(&pat_c, "%.*s", len_w, hun_y);
+  c3_assert( ret_i > 0 );
 
-    return pro;
+  c3_free(hun_y);
+
+  //  XX trim pat_c ?
+  //
+  return pat_c;
+}
+
+/* _king_get_next(): get next vere version string, if it exists.
+*/
+static c3_i
+_king_get_next(c3_c* pac_c, c3_c** out_c)
+{
+  c3_c* ver_c;
+  c3_c* url_c;
+  c3_w  len_w;
+  c3_y* hun_y;
+  c3_i  ret_i;
+
+  ret_i = asprintf(&url_c, "%s/%s/%s/next", ver_hos_c, pac_c, URBIT_VERSION);
+  c3_assert( ret_i > 0 );
+
+  if ( !_king_curl_bytes(url_c, &len_w, &hun_y) ) {
```

Comment:
Can you use the explicit numerical check instead?
```
  if ( 0 == _king_curl_bytes(url_c, &len_w, &hun_y) ) {
```

Reply:
In the past I might've agreed with you, but, for better or worse, this is idiomatic C for apis with return codes where `0` signals success.